### DISCUSSION TITLE
Improve documentation on ORM Mapper and Foreign Keys

### DISCUSSION THREAD

**Discussion by PythonCoderAS** [sqlalchemy/sqlalchemy#10724] [Ideas]
It is currently very confusing trying to figure out how to use the mapper and relationships, as none of the sample relationships code shows how to interact with the mapper. It took a lot of trial and error to learn that you need to make the `*_id` columns yourself and then pass them to the the `relationship()` call with the `foreign_keys` argument, like this:

```py
    thing_id: Mapped[int] = mapped_column(sqlalchemy.ForeignKey(Thing.id), nullable=False, init=False, repr=False)
    thing: Mapped[Thing] = sqlalchemy.orm.relationship(Thing, lazy="joined", innerjoin=True, foreign_keys=[thing_id], back_populates="other_things")
```

Now I can know I can set unique on the `thing_id` column instead of being confused as to why I can't set unique on `relationship()`.

Additionally, it would also help to know that in order to use many-to-many relationships a second class is basically required as SqlAlchemy will not autogenerate this table for you.

**Comment by zzzeek** [comment]
hey there -

docs are at https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html, you can literally copy and paste the code samples and they will run (we test them).    What docs are you referencing?

**Reply by zzzeek** [reply]
the "foreign_keys" relationship argument is a special case that is not usually needed

**Reply by PythonCoderAS** [reply]
I did many a Google search and somehow never found that page whatsoever. Maybe I just wasn't using the right keywords.

**Reply by CaselIT** [reply]
Seems that in this instance the sphinx search does find that page https://docs.sqlalchemy.org/en/20/search.html?q=relationship+&check_keywords=yes&area=default
